rust/parquet/src/file/serialized_reader.rs

```diff
@@ -137,6 +137,22 @@ impl<R: 'static + ChunkReader> SerializedFileReader<R> {
             metadata,
         })
     }
+
+    pub fn filter_row_groups(
```

Comment:
this is a fancy way to filter out the row groups -- it is probably worth adding documentation here.

I don't know if there are assumptions in the parquet reader code that the row group metadata matches what was read from the file or not

I suggest you consider filtering the row groups at the DataFusion (aka skip them in the datafusion physical operator) level rather than in the parquet reader level and avoid that potential problem completely.

Reply:
Yeah I think we can either move this to the application layer (i.e., data fusion), or expose it as a util function from `footer.rs`.

Reply:
Good point about documentation - will add some shortly. 

As long as row group metadata is filtered immediately after creating a SerializedFileReader, this approach will work.

That's the simplest way I could think of to allow filtering of row groups using statistics metadata; not sure how this could be done within DataFusion itself, because it reads data in batches (of configurable size) which could potentially span multiple row groups; it could be done, but would probably move a lot of complexity into DataFusion which today is nicely abstracted into the parquet library. This would also expose a lot more about the internals of a parquet file format to the outside as the user would have to be aware of row groups rather than just requesting batches of data.
May be I misunderstand what you are suggesting?

Reply:
there is another possibility - I have just noticed `FilePageIterator::with_row_groups` which could be used to filter row groups based on a list of row group indexes; this could replace the `filter_row_groups` method but would require the row group indexes to be passed down all the way to `build_for_primitive_type_inner` where `FilePageIterator` is created; this could be done through a new field in `ArrayReaderBuilderContext`.
It's a deeper change but would mean that `filter_row_groups` method is no longer necessary. @sunchao  do you think this would be a better way to go about filtering of row groups?

Reply:
What I was thinking is that, we can have another constructor for `SerializedFileReader` which takes a custom metadata:
```rust
    pub fn new_with_metadata(chunk_reader: R, metadata: ParquetMetaData) -> Result<Self> {
        Ok(Self {
            chunk_reader: Arc::new(chunk_reader),
            metadata: metadata,
        })
    }
```
and we move the metadata filtering part to data fusion, or a util function in `footer.rs`.

In the long term though, I think we should do something similar to [parquet-mr is doing](https://github.com/apache/parquet-mr/blob/master/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetFileReader.java#L656), that is, having a `ParquetReadOptions`-like struct which allows user to specify various configs, properties, filters, etc when reading a parquet file. The struct is extendable as well to accommodate new features in future such as filtering with column indexes or bloom filters. The constructor can become like this:
```rust
    pub fn new(chunk_reader: R, options: ParquetReadOptions) -> Result<Self> {
        Ok(Self {
            chunk_reader: Arc::new(chunk_reader),
            options: options,
        })
    }
```

Reply:
the second option, with the `ParquetReadOptions` parameter, sounds better - more extensible as you have described; however I think this falls outside of the scope of this PR

Reply:
> however I think this falls outside of the scope of this PR;

I agree -- this is already a large enough PR (and important enough). If we need to add some non ideal api to parquet and then upgrade it later I think that is the better approach.

Reply:
Yeah I didn't mean we should tackle it here - which is why I said "in the long term" :-)